libs/ml/include/ml/model/model.hpp

```diff
@@ -259,11 +298,33 @@ void Model<TensorType>::Predict(TensorType &input, TensorType &output)
 }
 
 template <typename TensorType>
-typename Model<TensorType>::DataType Model<TensorType>::Evaluate(
-    std::vector<MetricType> const &metrics)
+typename Model<TensorType>::DataVectorType Model<TensorType>::Evaluate(
+    dataloaders::DataLoaderMode dl_mode, SizeType batch_size)
 {
-  FETCH_UNUSED(metrics);
-  return loss_;
+  if (!compiled_)
+  {
+    throw ml::exceptions::InvalidMode("must compile model before evaluating");
+  }
+
+  dataloader_ptr_->SetMode(dl_mode);
+  bool is_done_set;
+  if (batch_size == 0)
+  {
+    batch_size = dataloader_ptr_->Size();
+  }
+  auto evaluate_pair = dataloader_ptr_->PrepareBatch(batch_size, is_done_set);
+
+  this->graph_ptr_->SetInput(label_, evaluate_pair.first);
+  this->graph_ptr_->SetInput(input_, evaluate_pair.second.at(0));
+
+  DataVectorType ret;
```

Comment:
a comment would make these subsequent lines less cryptic, i.e.:
"call evaluate on the graph and store the loss in ret", then
"push further metrics into ret - due to graph caching these subsequent evaluate calls are cheap"